### PR TITLE
Admin Router: Nested `/service` paths support

### PR DESCRIPTION
**Pull request by vespian**
## High Level Description

This PR enables nested service paths handling in for `/service` endpoint, i.e. `/service/group1/group2/jenkins`.

## Related Issues

https://jira.mesosphere.com/browse/DCOS-14891 - Admin Router paths for nested services

## Related PRs:

EE DC/OS PR:  https://github.com/mesosphere/dcos-enterprise/pull/929

### CODE REVIEW

File: packages/adminrouter/extra/src/lib/util.lua [hunk 1]
```diff
@@ -62,6 +62,56 @@ function util.table_len(tbl)
...
+
+    for i=1, math.floor(tl / 2) do
+        local tmp = tbl[i]
+        tbl[i] = tbl[tl - i + 1]
+        tbl[tl - i + 1] = tmp
+    end
+end
+
+
+function util.normalize_service_name(serviceName, fieldsLimit)
```
Comment (mhrabovcin) [inline packages/adminrouter/extra/src/lib/util.lua]: Can we add some documentation here? This seems like an important function.

Comment (vespian) [reply packages/adminrouter/extra/src/lib/util.lua]: Documentation added, here and in few other places. Please have another look!

File: packages/adminrouter/extra/src/lib/util.lua [hunk 2]
```diff
@@ -56,12 +56,116 @@ end
...
+    -- Extract path component in normalized form, at given level.
+    --
+    -- This function extracts from given service path the normalized service id
+    -- at a given level, defined by fieldsLimit.For example:
+    --
+    -- service_path group1/jenkins/ver/important/path/foobar.js
+    --
+    -- will yield following results, depending on fieldsLimit parameter:
+    -- nil:
+    --   normalised_name: foobar.js.path.important.ver.jenkins.group1
```
Comment (mhrabovcin) [inline packages/adminrouter/extra/src/lib/util.lua:90]: Should this be `js.foobar....` ?

File: packages/adminrouter/extra/src/lib/util.lua [hunk 3]
```diff
@@ -56,12 +56,116 @@ end
...
+    --   plain_name: group1/jenkins/ver/important/path/foobar.js
+    --   noreSegments: false
```
Comment (mhrabovcin) [inline packages/adminrouter/extra/src/lib/util.lua]: Typo `moreSegements`.

File: packages/adminrouter/extra/src/lib/util.lua (hunk 2 above)
Comment (mhrabovcin) [reply packages/adminrouter/extra/src/lib/util.lua:90]: Ah, I see that `foobar.js` is a single segment.

Comment (vespian) [reply packages/adminrouter/extra/src/lib/util.lua:90]: no, please notice that it's the last component of the path - `foobar.js`. The "inversion" happens on `/`.

Comment (vespian) [reply packages/adminrouter/extra/src/lib/util.lua:90]: sorry, did not notice your later comment :)

**Review by mhrabovcin — APPROVED** [review]

**Review by mhrabovcin — APPROVED** [review]

File: packages/adminrouter/extra/src/docs/api/nginx.master.yaml [hunk 4]
```diff
@@ -235,6 +232,17 @@ routes:
       path: 'http://$backend/lashup/key'
       backend: navstar
     path: /navstar/lashup/key
+  /exhibitor/exhibitor/v1/cluster/status:
+    group: Other
+    matcher: exact
+    proxy:
+      path: 'http://$backend= /exhibitor/exhibitor/v1/cluster/status'
```
Comment (gpaul) [inline packages/adminrouter/extra/src/docs/api/nginx.master.yaml]: This `$backend= ` looks a bit weird, is this correct?

File: packages/adminrouter/extra/src/docs/api/nginx.master.yaml [hunk 5]
```diff
@@ -264,6 +272,13 @@ routes:
...
     file: /opt/mesosphere/active.buildinfo.full.json
     path: /pkgpanda/active.buildinfo.full.json
+  /service/(?<service_path>.+):
+    group: Service
+    matcher: regex
+    description: Proxy to services running on DC/OS
+    proxy:
+      path: $upstream_url
+    path: /service/(?<service_path>.+)
   '/service/(?<serviceid>[0-9a-zA-Z-.]+)':
```
Comment (gpaul) [inline packages/adminrouter/extra/src/docs/api/nginx.master.yaml:283]: We should probably remove this now that it is superceded by your new code.

File: packages/adminrouter/extra/src/lib/util.lua [hunk 6]
```diff
@@ -56,12 +56,116 @@ end
...
+    --     - moreSegments: true/false depending on whether all segments
+    --       service_path were processed or not.
+    local tmpTbl = {}
+    local moreSegments = false
+
+    if service_path:len() > 0 then
+        fieldsLimit = fieldsLimit or -1
+
+        local fieldCursor, searchCursor = 1, 1
+        local substrStart, substrEnd = service_path:find('/', searchCursor, true)
```
Comment (gpaul) [inline packages/adminrouter/extra/src/lib/util.lua]: I believe as you're searching for a single-character string, `substrStart === substrEnd`. Perhaps we should use that fact and do
```
sepIdx, _ = service_path.find('/', ...)
```

File: packages/adminrouter/extra/src/lib/util.lua [hunk 7]
```diff
@@ -56,12 +56,116 @@ end
...
+
+    for i=1, math.floor(tl / 2) do
+        local tmp = tbl[i]
+        tbl[i] = tbl[tl - i + 1]
+        tbl[tl - i + 1] = tmp
+    end
+end
+
+
+function util.extract_service_path_component(service_path, fieldsLimit)
```
Comment (gpaul) [inline packages/adminrouter/extra/src/lib/util.lua:80]: lovely algorithm

**Review by gpaul — APPROVED** [review]
LGTM with one ignorable suggestion

File: packages/adminrouter/extra/src/lib/util.lua (hunk 3 above)
Comment (vespian) [reply packages/adminrouter/extra/src/lib/util.lua]: fixed!

File: packages/adminrouter/extra/src/docs/api/nginx.master.yaml (hunk 4 above)
Comment (vespian) [reply packages/adminrouter/extra/src/docs/api/nginx.master.yaml]: Nope, fixed!

File: packages/adminrouter/extra/src/docs/api/nginx.master.yaml (hunk 5 above)
Comment (vespian) [reply packages/adminrouter/extra/src/docs/api/nginx.master.yaml:283]: It's not - this entry describes a redirect location. Check the description below this line.

File: packages/adminrouter/extra/src/lib/util.lua (hunk 6 above)
Comment (vespian) [reply packages/adminrouter/extra/src/lib/util.lua]: Good catch! Fixing!

**Comment by vespian** [comment]
Merging!

**Pull request merged**